accepted/future-releases/static-extension-methods/feature-specification.md

```diff
@@ -46,214 +46,205 @@ extension MyFancyList<T> on List<T> {
+  - `E<...>(e)[e1]++` is equivalent to `E<...>(e)[e2] = E<...>(e)[e2] + 1` except that `e` and `e2` are only evaluated once, and the value of the increment expression is the value of the subexpression `E<...>(e)[e2]` before the addition. Symmetrically for post decrement.
 
-### Generic Parameter Inference
+### Implicit Extension Member Invocation
 
-If both the extension and the method is generic, then inference must infer the extension type parameters first, to figure out whether the extension applies, and only then start inferring method type parameters. As mentioned above, the inference is similar to other cases of chained inference. 
+Extension members can be invoked *implicitly* (without mentioning the extension by name) as if they were members of the `on` type of the extension. This is intended as the primary way to use extensions, with explicit extension member invocation as a fallback for cases where the implicit extension resolution doesn't do what the user want. 
 
-Example:
+An implicit extension member invocation occurs for a simple or composite member invocation with a target expression `e` iff there exists a unique *most specific* extension declaration which is *accessible* and *applicable* to the member invocation (see below).
 
-```dart
-extension SuperList<T> on List<T> {
-  R foldRight<R>(R base, R combine(T element, R accumulator)) {
-    for (int i = this.length - 1; i >= 0; i--) {
-      base = combine(this[i], base);
-    }
-    return base;
-  }
-}
-...
-  List<String> strings = ...;
-  int count(String string, int length) => length + string.length;
-  ...
-  var length = strings.foldRight(0, count);
-```
+If `E` is the single most specific accessible and applicable extension for a member invocation *i* with target expression `e`, which would then be a compile-time error by itself,  then we treat the target expression as if it was the extension application of `E` to `e`, and if `E` is generic, also providing the type arguments inferred for `E` in checking that it was applicable. This makes the member invocation equivalent to an explicit extension member invocation. This happens even if the *name* `E` is not accessible, so this is not a syntactic rewrite.
 
-Here the inference occurs just as if the extension had been declared like:
+Implicit extension member invocation can also apply to individual *cascade* invocations. A cascade is treated as if each cascade section was a separate member invocation on an expression with the same value as the cascade receiver expression (the one before the first `..`). This means that a cascade like `o..foo()..bar()` may perform an implicit extension member invocation on `o` for `foo()` and a normal invocation on `o` for `bar()`. There is no way to specify the corresponding explicit member invocation, so this is again not a syntactic rewrite.
```

Comment:
Note that cascades are themselves specified in terms of a syntactic rewrite (`evaluate t.suffix to an object`), and the rewrite would give rise to an extension invocation using the rules for normal member access. So there's no need to talk about syntactic rewrites here.

Reply:
Rewritten.